File: add_two_numbers.js
// You are given two non-empty linked lists representing two non-negative integers. 
// The digits are stored in reverse order and each of their nodes contain a single digit. 
// Add the two numbers and return it as a linked list.

// You may assume the two numbers do not contain any leading zero, except the number 0 itself.

// Example:

// Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
// Output: 7 -> 0 -> 8
// Explanation: 342 + 465 = 807.


function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

var addTwoNumbers = (l1, l2) => {
    var nbr = (reversed_list_int_representation(l1) +
              reversed_list_int_representation(l2));

    return (int_to_reversed_list(nbr));
}

const reversed_list_int_representation = lst => {
  var nbr_arr = [], 
      node = lst;

  while (node) {
    nbr_arr.push(node.val);
    node = node.next;
  }
  
  var result = nbr_arr.reverse().toString().replace(/,/g,'');
  return parseInt(result, 10);
}

const int_to_reversed_list = n => {
  n = n.toString().split('').reverse();  
  var nodes = [];
  
  for (let v of n) {
    nodes.push(new ListNode(v));
  }
  for (let i = 0; i < nodes.length-1; i++) {
    nodes[i].next = nodes[i+1];
  }

  return nodes[0];
}

{
let l1 = new ListNode(2);
let l2 = new ListNode(4);
let l3 = new ListNode(3);

l1.next = l2, l2.next = l3;

let l4 = new ListNode(5);
let l5 = new ListNode(6);
let l6 = new ListNode(4);

l4.next = l5, l5.next = l6;
// console.log(reversed_list_int_representation(l1));
// console.log(reversed_list_int_representation(l4));
// console.log(int_to_reversed_list(807));

console.log(addTwoNumbers(l1, l4));

}
/* 
  1. Function to turn Linked list representation of a reversed number into an Integer.
  2. Function to turn Integer into reversed linked list number representation.
*/

File: add_two_numbers.py
class ListNode():
    def __init__(self, val):
        self.val = val
        self.next = None

def nbr_to_list(nbr):
    res = [int(x) for x in str(nbr)]
    nodes = []
    for i in range(0,len(res)):
        nodes.append(ListNode(res[i])) 
    for i in range(0,len(nodes)-1):
        nodes[i].next = nodes[i+1]

    return (nodes[0])

def reversed_list_to_nbr(lst):
    nbr = 0
    while (lst):
        nbr += lst.val
        nbr *= 10
        lst = lst.next
    nbr = str(nbr)
    nbr = nbr[0:-1]
    nbr = nbr[::-1]
    return (int(nbr))

def nbr_to_reversed_list(nbr):
    nbr = str(nbr)[::-1]
    return nbr_to_list(int(nbr))


def add_two_numbers(l1, l2):
    n1 = reversed_list_to_nbr(l1)
    n2 = reversed_list_to_nbr(l2)
    
    res_head = nbr_to_reversed_list(n1+n2)
    return (res_head)

l1 = nbr_to_list(243) 
l2 = nbr_to_list(564)
result = add_two_numbers(l1, l2)
print(result.next.next.val)

# print(reversed_list_to_nbr(l1))
# print(reversed_list_to_nbr(l2))

# for x in nodes:
#     print("val: "+str(x.val)+" next: "+str(x.next))








#****************************************************
#****************************************************
#****************************************************
#****************************************************
class Solution(object):
    def addTwoNumbers(self,l1, l2):
        n1 = self.reversed_list_to_nbr(l1)
        n2 = self.reversed_list_to_nbr(l2)

        res_head = self.nbr_to_reversed_list(n1+n2)
        return (res_head)
    
    def nbr_to_list(self,nbr):
        res = [int(x) for x in str(nbr)]
        nodes = []
        for i in range(0,len(res)):
            nodes.append(ListNode(res[i])) 
        for i in range(0,len(nodes)-1):
            nodes[i].next = nodes[i+1]

        return (nodes[0])

    def reversed_list_to_nbr(self,lst):
        nbr = 0
        while (lst):
            nbr += lst.val
            nbr *= 10
            lst = lst.next
        nbr = str(nbr)
        nbr = nbr[0:-1]
        nbr = nbr[::-1]
        return (int(nbr))

    def nbr_to_reversed_list(self,nbr):
        nbr = str(nbr)[::-1]
        return self.nbr_to_list(int(nbr))